Fix support for services connected via IPv6

edomora97 opened the pull request:
Use the correct address family when connecting to the various RPC services.

The solution is basically the same suggested here:
https://github.com/cms-dev/cms/issues/157#issuecomment-463427954

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cms-dev/cms/1141)
<!-- Reviewable:end -->

andreyv commented on cms/io/rpc.py:
```diff
@@ -478,8 +479,12 @@ def _connect(self):
+            # Find the family (AF_INET or AF_INET6) of the IPv4/6 address.
+            addresses = gevent.socket.getaddrinfo(
+                self.remote_address.ip, self.remote_address.port)
+            family, *rest, sockaddr = addresses[0]
```
It is slightly worrying to take only the first returned address. It would be more prudent to try all of them. From Linux `getaddrinfo(3)`:
> Normally, the application should try using the addresses in the order in which they are returned.

Can you please implement a loop that tries all addresses in succession? A related example is available at https://docs.python.org/3/library/socket.html#example.

edomora97 replied:
Done, tell me if you like the loop like that.

andreyv commented on cms/io/service.py:
```diff
@@ -158,9 +159,11 @@ def _connection_handler(self, sock, address):
-            ipaddr = gevent.socket.gethostbyname(ipaddr)
-            address = Address(ipaddr, port)
+            # In case of IPv6 address is (ip, port, flow info, scope id).
+            ipaddr, port = address[:2]
```
In this place, the address is ultimately coming from `accept()`, and should have the numerical form already. So I think we can just create an `Address` directly from that without any address resolution.

edomora97 replied:
Right, fixed.

andreyv reviewed (changes requested):
Thanks! Overall this looks good. Please see comments below.

edomora97 commented:
Sure! I'll take a look again at this in few days, I'm pretty busy right now!

edomora97 commented on cms/io/rpc.py:
```diff
@@ -478,8 +479,12 @@ def _connect(self):
-            sock = gevent.socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            sock.connect(self.remote_address)
+            # Find the family (AF_INET or AF_INET6) of the IPv4/6 address.
+            addresses = gevent.socket.getaddrinfo(
```
Done.

edomora97 commented:
Sorry for the _small_ delay, I've rebased on the current master and fixed the comments you left. The first two commits are the exact same, the third fixes the code review (hopefully).

andreyv commented on cms/io/rpc.py:
```diff
@@ -478,13 +479,27 @@ def _connect(self):
+                logger.debug("Couldn't connect to %s at %s port %d: %s.",
+                             self._repr_remote(), host, port, error)
+            else:
+                self.initialize(sock, self.remote_service_coord)
```
Is a `break` missing here?

edomora97 replied:
Yes. Fixing it immediately.

andreyv reviewed (changes requested):
Hi,

Thanks for the update. Now, re-reading the changes, I started wondering: do we need `getaddrinfo()` at all? [`Address`](https://github.com/cms-dev/cms/blob/da7f2d26b821f4f90f24201adc920ab1ad4ced61/cms/conf.py#L37) already contains an IP address, so do we need to resolve it again?

It looks like it should be enough to just get `ipaddress.ip_address(address[0]).version` and select address family based on that. What do you think?

edomora97 commented:
I don't think `ipaddress.ip_address(address[0]).version` is enough. That `Address` can also contain domain names (as you write them in `cms.conf`), and `ipaddress.ip_address` only parses plain IP addresses.

andreyv commented on cms/io/rpc.py:
```diff
@@ -478,13 +479,28 @@ def _connect(self):
+            logger.warning("Cannot resolve %s.", self.remote_address)
+            raise
+
+        for family, *rest, sockaddr in addresses:
```
Sorry, one more comment: you can get `type` and `proto` here and pass them to `socket()`, as recommended in the docs.

andreyv commented:
Thanks, the code looks good now and works fine with both IPv4 and IPv6.

andreyv commented:
I rebased this branch onto de5a0caf9b9ae02e4a90618748cbb46a66189b11.

One unit test fails, mock description needs to be adjusted:
```
AssertionError: expected call not found.
Expected: getaddrinfo('127.0.0.1', 36887)
Actual: getaddrinfo('127.0.0.1', 36887, type=<SocketKind.SOCK_STREAM: 1>)
```

andreyv commented:
Looks good, thanks.
